### PR TITLE
Fix cloneBlockchainTime

### LINKED ISSUE #1489: Fix cloning of `TestBlockChaintime`
PR #1202 introduced a bug in the test infrastructure that prevents nodes from forging during ~slot 0~ the slot in which they join.

The new `BlockhainTime` cloning introduced there does not allow the cloned `btime`'s `onSlotChange` from ticking with `s` as its argument if `s` is the slot in which the node joins the net.

This Issue is to fix that.

### LINKED ISSUE #1524: Test with `k = 2` in test-consensus
Like Issue #1511 (k = 1). I've observed one failure in RealPBFT at `k = 2`.

This Issue is to either fix `Network.hs` so `k = 2` doesn't cause failures like this or to fix `RealPBFT.hs` to not generate cases that fail in this way; preferably the former.

This is the repro:

```
      prop_simple_real_pbft_convergence 
        ProduceEBBs
        SecurityParam {maxRollbacks = 2}   -- boundaries at 0 20
        TestConfig
          { numCoreNodes = NumCoreNodes 5
          , numSlots     = NumSlots 25
          , nodeJoinPlan = NodeJoinPlan $ Map.fromList
              [ (CoreNodeId 0, SlotNo 0)
              , (CoreNodeId 1, SlotNo 0)
              , (CoreNodeId 2, SlotNo 0)
              , (CoreNodeId 3, SlotNo 23)
              , (CoreNodeId 4, SlotNo 23)
              ]
          , nodeRestarts = noRestarts
          , nodeTopology = NodeTopology $ Map.fromList
              -- linear
              [ (CoreNodeId 0, Set.fromList [])
              , (CoreNodeId 1, Set.fromList [CoreNodeId 0])
              , (CoreNodeId 2, Set.fromList [CoreNodeId 1])
              , (CoreNodeId 3, Set.fromList [CoreNodeId 2])
              , (CoreNodeId 4, Set.fromList [CoreNodeId 3])
              ]
          , slotLengths  = ...
          , initSeed     = ...
          }
```

The failure was: lack of consensus in final chains. The competing final chain suffixes look like this:

```
c0-c3 have
( header: ( hash: e2b8f47b27814ce8, slot: 16, issuer: c1)) :>
( header: ( hash: 7a24f7153e9fe779, slot: 17, issuer: c2)) :>
( ebb: 1,   hash: 9f69106d3569fe30, slot: 20) :>
( header: ( hash: 49b15cd694a2acd9, slot: 20, issuer: c0)) :>
( header: ( hash: d7a83022abcba3f9, slot: 21, issuer: c1)) :>
( header: ( hash: 223e857131ea40d4, slot: 22, issuer: c2))

c4 has
( header: ( hash: e2b8f47b27814ce8, slot: 16, issuer: c1)) :>
( ebb: 1,   hash: ccbfdb464c7a79f6, slot: 20) :>
( header: ( hash: f689d420a9811cd7, slot: 23, issuer: c3)) :>
( header: ( hash: 074eb97a7ca1a4a3, slot: 24, issuer: c4))
```

A lot happens in slot s23:

```
...
(SlotNo {unSlotNo = 23},(CoreNodeId 2,CoreNodeId 3),MiniProtocolDelayed,MPEEChainSyncClient (ForkTooDeep {_intersection = Origin, _ourTip = Tip {tipPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 22}, blockPointHash = ByronHash {unByronHash = AbstractHash 223e857131ea40d41c332e2ff412f03633c90e5ceea9b4bdf4b82f04ff1fb5f8}}), tipBlockNo = BlockNo {unBlockNo = 15}}, _theirTip = Tip {tipPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 0}, blockPointHash = ByronHash {unByronHash = AbstractHash 3c2a8234e7b7b68b749b9ffd2a5ca4b8dbd97d4ae08285524f109abaa967f980}}), tipBlockNo = BlockNo {unBlockNo = 0}}}))
...
c3 SwitchedToChain ... s16  e2b8f47b
... c3 forges EBB ccbfdb46 ...
c3 SwitchedToChain ... s20  ccbfdb46
... c3 forges f689d420 ...
c3 SwitchedToChain ... s23  f689d420
(SlotNo 23,(CoreNodeId 3,CoreNodeId 4),MiniProtocolDelayed,MPEEChainSyncClient (InvalidRollForward
  { _newPoint =                 At (BlockPoint s16 e2b8f47b)

  -- this _ourTip field is out-of-date; it's from the latest time we found the intersection;
  -- the hypothetical _ourCurrentTip would instead be s23 f689d420 b12
  -- (or maybe s20 9f69106d b12?)

  , _ourTip   = Tip {tipPoint = At (BlockPoint s15 52232fc6), tipBlockNo = BlockNo 10}

  -- this _theirTip field, on the otherhand, is up-to-date, from the
  -- RollForward message itself

  , _theirTip = Tip {tipPoint = At (BlockPoint s16 e2b8f47b), tipBlockNo = BlockNo 11}
  }))
c3 SwitchedToChain ... s20  9f69106d   -- c0 forged this EBB
c3 SwitchedToChain ... s20  49b15cd6   -- c0 forged this proper block
c3 SwitchedToChain ... s21  d7a83022   -- c1 forged this block
(SlotNo 23,(CoreNodeId 4,CoreNodeId 3),MiniProtocolDelayed,MPEEChainSyncClient (InvalidRollForward
  { _newPoint =                 At (BlockPoint s17 7a24f715)
  , _ourTip   = Tip {tipPoint = At (BlockPoint s23 f689d420), tipBlockNo = BlockNo 12}
  , _theirTip = Tip {tipPoint = At (BlockPoint s20 9f69106d), tipBlockNo = BlockNo 12}
  }))
c3 SwitchedToChain ... s22  223e8571   -- c2 forged this block
```

I think the mini protocol failure for client=c4 and server=c3 (ie `(CoreNodeId 4,CoreNodeId 3)`) is the crux of the issue. And I think it ultimately happens because the chain c3 forges violates the `at least k blocks in 2k slots` invariant. Lastly, I think that that happens because the `forkEbbProducer` thread in `Network.hs` uses `>=` instead of `>` when simulating the `anachronisticLedgerView`'s `TooFarAhead` check.

In slot s24, the mini protocol failures culminate in a wedge in the chains, which directly causes the test failure:

```
(SlotNo {unSlotNo = 24},(CoreNodeId 2,CoreNodeId 3),MiniProtocolRestarting,MPEEChainSyncClient (ForkTooDeep {_intersection = Origin, _ourTip = Tip {tipPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 22}, blockPointHash = ByronHash {unByronHash = AbstractHash 223e857131ea40d41c332e2ff412f03633c90e5ceea9b4bdf4b82f04ff1fb5f8}}), tipBlockNo = BlockNo {unBlockNo = 15}}, _theirTip = Tip {tipPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 0}, blockPointHash = ByronHash {unByronHash = AbstractHash 3c2a8234e7b7b68b749b9ffd2a5ca4b8dbd97d4ae08285524f109abaa967f980}}), tipBlockNo = BlockNo {unBlockNo = 0}}}))
(SlotNo {unSlotNo = 24},(CoreNodeId 3,CoreNodeId 4),MiniProtocolRestarting,MPEEChainSyncClient (InvalidRollForward {_newPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 16}, blockPointHash = ByronHash {unByronHash = AbstractHash e2b8f47b27814ce82e1e0589d4a80b9156d5bb06b4d33f57a442ded3b5eac23b}}), _ourTip = Tip {tipPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 15}, blockPointHash = ByronHash {unByronHash = AbstractHash 52232fc61f2184befb3bbaeda44631f5a7ee582c6338fff4b8d860055eb1d94d}}), tipBlockNo = BlockNo {unBlockNo = 10}}, _theirTip = Tip {tipPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 16}, blockPointHash = ByronHash {unByronHash = AbstractHash e2b8f47b27814ce82e1e0589d4a80b9156d5bb06b4d33f57a442ded3b5eac23b}}), tipBlockNo = BlockNo {unBlockNo = 11}}}))
(SlotNo {unSlotNo = 24},(CoreNodeId 4,CoreNodeId 3),MiniProtocolRestarting,MPEEChainSyncClient (InvalidRollForward {_newPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 17}, blockPointHash = ByronHash {unByronHash = AbstractHash 7a24f7153e9fe77943ffc597cdb83c9cdc834f3352ab9ac0f4ff2c5272f3f9e4}}), _ourTip = Tip {tipPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 23}, blockPointHash = ByronHash {unByronHash = AbstractHash f689d420a9811cd7de0a37a6015a576985574928a643e5d3233f99c89fc4df92}}), tipBlockNo = BlockNo {unBlockNo = 12}}, _theirTip = Tip {tipPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 20}, blockPointHash = ByronHash {unByronHash = AbstractHash 9f69106d3569fe30bd2bc8138fe5175c09660314285981e682326d7dcdfff337}}), tipBlockNo = BlockNo {unBlockNo = 12}}}))
... c4 forges 074eb97a7ca1a4a3 ...
(SlotNo {unSlotNo = 24},(CoreNodeId 3,CoreNodeId 4),MiniProtocolDelayed,MPEEChainSyncClient (ForkTooDeep {_intersection = Origin, _ourTip = Tip {tipPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 22}, blockPointHash = ByronHash {unByronHash = AbstractHash 223e857131ea40d41c332e2ff412f03633c90e5ceea9b4bdf4b82f04ff1fb5f8}}), tipBlockNo = BlockNo {unBlockNo = 15}}, _theirTip = Tip {tipPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 24}, blockPointHash = ByronHash {unByronHash = AbstractHash 074eb97a7ca1a4a3155b63f9dcd94712a56e0b357421761d75fba1ad341bc4b1}}), tipBlockNo = BlockNo {unBlockNo = 13}}}))
(SlotNo {unSlotNo = 24},(CoreNodeId 4,CoreNodeId 3),MiniProtocolDelayed,MPEEChainSyncClient (NoMoreIntersection {_ourTip = Tip {tipPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 24}, blockPointHash = ByronHash {unByronHash = AbstractHash 074eb97a7ca1a4a3155b63f9dcd94712a56e0b357421761d75fba1ad341bc4b1}}), tipBlockNo = BlockNo {unBlockNo = 13}}, _theirTip = Tip {tipPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 22}, blockPointHash = ByronHash {unByronHash = AbstractHash 223e857131ea40d41c332e2ff412f03633c90e5ceea9b4bdf4b82f04ff1fb5f8}}), tipBlockNo = BlockNo {unBlockNo = 15}}}))
```

### PR DESCRIPTION
**Pull request by nfrisby**
Fixes #1489. Fixes #1524.

Fixing Issue #1489 (let nodes lead when they join) and letting k vary in the
range [2 .. 10] since the dual-ledger tests do that now revealed several
Issues.

Issues in the library, not just the test infrastructure:

  * Issue #1505 -- `removeTxs` cannot use the fast path when validating after
    removing or else we might have dangling tx inputs references. Was fixed
    by #1565.

  * Issue #1511, bullet 1 (closed) -- The `Empty` cases in
    `prevPointAndBlockNo` were wrong. Recent PRs have addressed this: #1544 and
    #1589.

  * Issue #1543 (closed) -- A bracket in `registeredStream` was spoiled by an
    interruptible operation. Thomas' PR re-designed the vulnerability away. I
    think this is unrelated to the other changes; it was lurking and happened
    to pop up here just because I've been running hundreds of thousands of
    tests.

Issues only in the test infrastructure:

  * Issue #1489: let nodes lead when they join. This bug slipped in recently,
    when I added cloning of `BlockchainTime`s as part of the
    restarting/rekeying loop in the test infrastructure.

  * _PBFT reference simulator_. (Was masked by 1489.) Model competing 1-block
    chains in Ref.PBFT and use its results where applicable instead of the
    .Expectations module. Check that the PBFT threadnet and Ref.PBFT.simulate
    results agree on `Ref.Nominal` slots.

    The Ref.PBFT module had been making assumptions that were accurate given
    the guards in RealPBFT generators, given the `k >= n` regime. But outside
    of that regime, when the security parameter exceeds the node count, it
    wasn't enough. Also, it couldn't be compared against the PBFT threadnet
    because of those assumptions.

  * _PBFT reference simulator_. (Cascade of above.) Add
    `definitelyEnoughBlocks` to confirm the "at least k blocks in 2k slots"
    invariant in `genRealPBFTNodeJoinPlan`. The existing guards in the RealPBFT
    generators are intentionally insufficient by themselves; this way we can
    optimize them to avoid O(n^2) complexity without risking divergence from
    the `suchThat`s.

  * _Origin corner-case_. (Was masked by 1489.) Discard DualPBFT tests that
    forge in slot 0. The current `cardano-ledger-specs` doesn't allow for that.
    My hypothesis is that `cvsLastSlot` would need to be able to represent
    origin.

  * _Dlg cert tx_. Adjust `genNodeRekeys` to respect "If node X rekeys in slot
    S and Y leads slot S+1, then either the topology must connect X and Y
    directly, or Y must join before slot S."

  * _Dlg cert tx_. (Was (statistically?) masked by relatively large k.) Add
    `rekeyOracle` and use it to determine which epoch number to record in the
    dlg cert. The correct value depends on which block the dlg cert tx will end
    up in, not when we first add it to our mempool.

  * _Dlg cert tx_. (Was (statistically?) masked by relatively large k.) Add the
    dlg cert tx to the mempool each time the ledger changes.

### CODE REVIEW

File: ouroboros-consensus/test-consensus/Test/ThreadNet/BFT.hs [hunk 1]
```diff
@@ -35,6 +35,7 @@ prop_simple_bft_convergence k
     prop_general k
         testConfig
         (Just $ roundRobinLeaderSchedule numCoreNodes numSlots)
+        Nothing
```
Comment (nfrisby) [inline ouroboros-consensus/test-consensus/Test/ThreadNet/BFT.hs]: TODO cleanup the bare `Maybe` arguments to `prop_general`.

File: ouroboros-consensus/test-consensus/Test/ThreadNet/DualPBFT.hs [hunk 2]
```diff
@@ -45,12 +46,38 @@ tests = testGroup "DualPBFT" [
...
       (setupExpectedRejections setup)
       (setupTestOutput         setup)
+  where
+    cfg = setupConfig setup
+
+    refResult :: Ref.Result
+    refResult =
+      Ref.simulate (setupParams setup) (nodeJoinPlan cfg) (numSlots cfg)
+
+    -- The test infrastructure allows nodes to forge in slot 0; however, the
```
Comment (nfrisby) [inline ouroboros-consensus/test-consensus/Test/ThreadNet/DualPBFT.hs]: @edsko: does this make sense? At the low-level mechanics level, this restriction is "due" to `cvsLastSlot :: SlotNo` instead of `cvsLastSlot :: WithOrigin SlotNo`.

File: ouroboros-consensus/test-consensus/Test/ThreadNet/Network.hs [hunk 3]
```diff
@@ -376,16 +376,21 @@ runThreadNetwork ThreadNetworkArgs
...
               (NodeRekey, Just rekeyM) -> do
-                eno <- epochInfoEpoch epochInfo s
+                let ProtocolInfo{pInfoConfig} = pInfo
+                    k = protocolSecurityParam pInfoConfig
+                    -- the delegation will mature 2k slots after it is added to
+                    -- the chain
+                    --
+                    -- TODO we assume it's added in the current slot
+                    maturationSlot = s + 2 * SlotNo (maxRollbacks k)
+                eno <- epochInfoEpoch epochInfo maturationSlot
```
Comment (nfrisby) [inline ouroboros-consensus/test-consensus/Test/ThreadNet/Network.hs]: TODO Revisit this.

This change eliminated the error I was seeing and has passed ~600 (?) test, but I haven't found a similar `2k` shift in the certificate validation code in the `cardano-ledger` package.

File: ouroboros-consensus/test-consensus/Test/ThreadNet/Network.hs [hunk 4]
```diff
@@ -625,21 +654,30 @@ runThreadNetwork ThreadNetworkArgs
             , blockMatchesHeader  = nodeBlockMatchesHeader
             , maxUnackTxs         = 1000 -- TODO
             , maxBlockSize        = NoOverride
-            , mempoolCap          = NoMempoolCapacityBytesOverride
+            , mempoolCap          = MempoolCapacityBytesOverride $ MempoolCapacityBytes 3000
```
Comment (nfrisby) [inline ouroboros-consensus/test-consensus/Test/ThreadNet/Network.hs]: TODO This is a temporary workaround for Issue #1505 .

File: ouroboros-consensus/test-consensus/Test/ThreadNet/PBFT.hs [hunk 5]
```diff
@@ -1,45 +1,192 @@
...
+import           Test.ThreadNet.Util.NodeRestarts
+import           Test.ThreadNet.Util.NodeTopology
 
 import           Test.Util.Orphans.Arbitrary ()
 
+import Debug.Trace
+
 tests :: TestTree
 tests = testGroup "PBFT" [
-      testProperty "simple convergence" $
```
Comment (nfrisby) [inline ouroboros-consensus/test-consensus/Test/ThreadNet/PBFT.hs]: TODO Remove this accidentally committed code.

File: ouroboros-consensus/test-consensus/Test/ThreadNet/RealPBFT.hs [hunk 6]
```diff
@@ -194,9 +202,33 @@ tests = testGroup "RealPBFT" $
...
+          , nodeJoinPlan = trivialNodeJoinPlan ncn4
+          , nodeRestarts = NodeRestarts (Map.fromList [(SlotNo 59,Map.fromList [(CoreNodeId 3,NodeRekey)])])
+          , nodeTopology = meshNodeTopology ncn4
+          , slotLengths  = defaultSlotLengths
+          , initSeed     = Seed (17364222041321661634,8266509462575908621,10410472349244348261,9332246846568887555,6178891282750652496)
+          }
     , testProperty "simple convergence" $
           \produceEBBs ->
-          forAll (SecurityParam <$> elements [5, 10])
+          -- TODO k = 1 causes various errors
```
Comment (nfrisby) [inline ouroboros-consensus/test-consensus/Test/ThreadNet/RealPBFT.hs]: TODO Fix this.

@edsko: Would you confirm that `k = 1` is a valid parameter assignment?

File: ouroboros-consensus/test-consensus/Test/ThreadNet/Ref/PBFT.hs [hunk 7]
```diff
@@ -0,0 +1,620 @@
+{-# LANGUAGE LambdaCase          #-}
```
Comment (nfrisby) [inline ouroboros-consensus/test-consensus/Test/ThreadNet/Ref/PBFT.hs]: See the commits for a delta without the renaming (can't pass the `git diff -M` option to GitHub's PR renderer?).

File: ouroboros-consensus/test-consensus/Test/ThreadNet/RealPBFT.hs (hunk 6 above)
Comment (edsko) [reply ouroboros-consensus/test-consensus/Test/ThreadNet/RealPBFT.hs]: Yes. Even `k = 0` in _principle_ ought to be, but the Byron ledger spec doesn't like it. I _think_ the implementation ought to be able to handle it, but not totally sure about that. But I have seen no problems with `k = 1`.

File: ouroboros-consensus/test-consensus/Test/ThreadNet/DualPBFT.hs (hunk 2 above)
Comment (edsko) [reply ouroboros-consensus/test-consensus/Test/ThreadNet/DualPBFT.hs]: I see. I think it's okay for now. The treatment of the point near genesis is a mess to be honest, we ought to clean that up at some point.

**Comment by nfrisby** [comment]
I've pushed new commits and updated the PR description, but the commit history is still just a trail of breadcrumbs along my bug hunt. When I'm back online in about 6 hours, my immediate goal is to clean up those commits. The tip of this branch has passed 50000 RealPBFT k=2 tests and 50000 RealPBFT k=3 tests and tens of thousands of more varying tests.

**Comment by nfrisby** [comment]
That rebase cleaned up the git history, removing a lot of the code that arose during debugging.

`test-consensus: let k vary in PBFT and RealPBFT tests` b8d8ab5 is still a pretty big commit, but I'm not sure that splitting it up (into up to 3 commits, I'd guess) is worth the effort.

File: ouroboros-consensus/test-consensus/Test/ThreadNet/RealPBFT.hs [hunk 8]
```diff
@@ -669,45 +1023,54 @@ shrinkTestConfigSlotsOnly TestConfig
...
+                  NodeJoinPlan $ Map.insert cid (jslot + twoK) njp
+          , Ref.definitelyEnoughBlocks params $
+            Ref.simulate params nodeJoinPlan' numSlots
+            -- ... and does not join in the same slot as the leader of the next
+            -- slot unless they are neighbors (otherwise the dlg cert tx might
+            -- not reach it soon enough)
+          , let nextLeader = Ref.mkLeaderOf params $ succ jslot
+          , jslot /= coreNodeIdJoinSlot nodeJoinPlan nextLeader ||
+            cid `elem` coreNodeIdNeighbors nodeTopology nextLeader
           -> pure $ NodeRestarts $
```
Comment (nfrisby) [inline ouroboros-consensus/test-consensus/Test/ThreadNet/RealPBFT.hs]: We also need to prevent the rekeying node from restarting too soon, since it might need to propagate its dlg cert tx more than once.

File: ouroboros-consensus/test-consensus/Test/ThreadNet/Network.hs (hunk 3 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/ThreadNet/Network.hs]: I revisited this. I changed it, and I think it's right now. I introduced an oracle so that the node would know in which slot the next block will be forged.

File: ouroboros-consensus/test-consensus/Test/ThreadNet/PBFT.hs (hunk 5 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/ThreadNet/PBFT.hs]: The accidental code is gone. Done.

File: ouroboros-consensus/test-consensus/Test/ThreadNet/RealPBFT.hs (hunk 6 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/ThreadNet/RealPBFT.hs]: I'm deferring this to Issue 1511 (let k = 1 in RealPBFT). That Issue discusses some of the concrete challenges.

**Comment by nfrisby** [comment]
That force push is passing tests. Remaining work:

  * resolve a few TODOs in the code and from this PR's comments
  * rebase to origin/master (blocked at least by cardano-ledger#707)

File: ouroboros-consensus/test-consensus/Test/ThreadNet/General.hs [hunk 9]
```diff
@@ -169,7 +169,11 @@ data TestConfigBlock blk = TestConfigBlock
   }
 
 data Rekeying blk = forall opKey. Rekeying
-  { rekeyUpd ::
+  { rekeyOracle
```
Comment (nfrisby) [inline ouroboros-consensus/test-consensus/Test/ThreadNet/General.hs]: TODO Can we simplify away this callback? Specifically, can we create the cert as part of the test-configuration (eg make `NodeRestarts` a container, with items as a new argument to the `NodeRekey` ctor)?

File: ouroboros-consensus/test-consensus/Test/ThreadNet/Network.hs (hunk 4 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/ThreadNet/Network.hs]: Replaced by a rough draft of the appropriate fix (ie Issue 1565).

**Comment by nfrisby** [comment]
I rebased onto cee384a1, because that's the last commit on `master` that still has the Byron ledger wrapper in this repo. My patch for `cardano-ledger` Issue 715 cannot exist in this repo beyond that commit.

Unfortunately, the fixes for 1544 (`prevPointAndBlockNo`) and 1565 (`removeTxs`) come later.

I'm going to open a PR in `cardano-ledger` if only to get the ball rolling.

**Comment by nfrisby** [comment]
I opened input-output-hk/cardano-ledger#716, which I think is the necessary fix (it is the same as the patch I have locally on this branch). Once that's merged, then my PR can update the deps in the repo accordingly and therefore rebase all the way to our `origin/master`.

**Comment by nfrisby** [comment]
After rebasing onto master (now that a `cardano-ledger` fix commit is available for gitfetch), I'm seeing errors that I believe are due to Issue #1147. I see the characteristic `FetchDeclineConcurrencyLimit` BF decision between a `CompletedBlockFetch` and `CompletedFetchBatch`, and if I cherry-pick my patch for that from PR 1131, then the failing repro passes.

I'm guessing the recent changes to ChainSync client (in light of the recent refinement of tips to possibly include "origin" instead of an off-by-one blockNumber = 0") has somehow enabled this family of mini protocol message/event interleavings, which PR 1131 instead has been exploring by adding network latencies.

cc: @dcoutts @edsko @mrBliss

**Comment by nfrisby** [comment]
I eventually realized that the Just-In-Time EBBs (enabled by a recent PR on master) avoid the error case that otherwise required a patch for Issue 1631 (Chain Density check in `NodeKernel`) and a rewrite of the reference simulator. This meant I could drop the `NodeKernel` change, which simplified the PR quite a bit. It no longer addresses 1631, leaving that for future work.

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/Pam.hs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+import qualified Data.Map as Map
+import           Data.Type.Coercion
+
+-- | An inverted 'Map'
+--
+-- INVARIANT the @k@s are all unique
+--
+-- INVARIANT the 'NonEmpty's are all ascending
+--
+newtype Pam v k = UnsafePam {getPam :: Map v (NonEmpty k)}
```
Comment (edsko) [inline ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/Pam.hs]: Let's rename this to `InvertedMap`.

File: ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime/Mock.hs [hunk 11]
```diff
@@ -74,8 +73,10 @@ data TestClock =
   deriving (Eq, Generic, NoUnexpectedThunks)
 
 data TestBlockchainTime m = TestBlockchainTime
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime/Mock.hs]: We really ought to move this to `test-infra`.

File: ouroboros-consensus-byron/test/Test/ThreadNet/RealPBFT.hs [hunk 12]
```diff
@@ -81,10 +83,18 @@ import           Test.Util.Orphans.Arbitrary ()
 import           Test.Util.Shrink (andId, dropId)
 import qualified Test.Util.Stream as Stream
 
+newtype K = K SecurityParam
+  deriving (Show)
+
+instance Arbitrary K where
```
Comment (edsko) [inline ouroboros-consensus-byron/test/Test/ThreadNet/RealPBFT.hs]: `SecurityParam` already has an `Arbitrary` instance. Maybe give an instance for `Small SecurityParam` instead? (And can then just call `shrink`).

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs [hunk 13]
```diff
@@ -116,25 +117,35 @@ import           Test.Util.Tracer
...
+         TopLevelConfig blk
+      -> SlotNo
+         -- EBB slot
+      -> BlockNo
+         -- EBB block number (i.e. that of its predecessor)
+      -> ChainHash blk
+         -- EBB predecessor's hash
+      -> blk
+  , ebbSlotBefore :: SlotNo -> SlotNo
+    -- ^ the slot of the most recent expected that precedes the given slot
```
Comment (edsko) [inline ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs]: I don't understand this name or this comment :)

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs [hunk 14]
```diff
@@ -428,18 +436,41 @@ runThreadNetwork ThreadNetworkArgs
...
             Nothing                     -> pure ()
             Just (s', pInfo', nr', rs') -> loop s' pInfo' nr' rs'
 
+    -- | Persistently attempt to add the given transactions to the mempool
+    -- every time the ledger slot changes, even if successful!
+    --
+    -- If we add the transaction and then the mempools discards it for some
+    -- reason, this thread will add it again.
+    --
+    forkTxs0
```
Comment (edsko) [inline ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs]: This could do with a one or two line explanation of what the purpose of it is.

File: ouroboros-consensus-byron/test/Test/ThreadNet/DualPBFT.hs [hunk 15]
```diff
@@ -72,13 +73,40 @@ tests = testGroup "DualPBFT" [
...
+  where
+    cfg = setupConfig setup
+
+    refResult :: Ref.Result
+    refResult =
+      Ref.simulate (setupParams setup) (nodeJoinPlan cfg) (numSlots cfg)
+
+    -- The test infrastructure allows nodes to forge in slot 0; however, the
+    -- cardano-ledger-specs code causes @PBFTFailure (SlotNotAfterLastBlock
+    -- (Slot 0) (Slot 0))@ in that case. So we discard such tests.
```
Comment (edsko) [inline ouroboros-consensus-byron/test/Test/ThreadNet/DualPBFT.hs]: Could do with a comment that the spec here is therefore more strict than the real implementation.

File: ouroboros-consensus-byron/test/Test/ThreadNet/RealPBFT.hs (hunk 12 above)
Comment (nfrisby) [reply ouroboros-consensus-byron/test/Test/ThreadNet/RealPBFT.hs]: Elaborated in commit 1acb1cbe2cebcc7afc8b829d9310e48fc85456c3

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs (hunk 13 above)
Comment (nfrisby) [reply ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs]: Elaborated in commit 9b2aa86f61b252315479113627e679953d02c542

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs (hunk 14 above)
Comment (nfrisby) [reply ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs]: Elaborated in commit 183da45e73336e41c25473b55a33099467cc75ce

File: ouroboros-consensus-byron/test/Test/ThreadNet/DualPBFT.hs (hunk 15 above)
Comment (nfrisby) [reply ouroboros-consensus-byron/test/Test/ThreadNet/DualPBFT.hs]: Elaborated in commit d0134cac7a0c684128b645903435d26412546c2f

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/Pam.hs (hunk 10 above)
Comment (nfrisby) [reply ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/Pam.hs]: Done in commit 3de1fc8401ba1a57fbe875fba9180217a1f98c92

**Comment by nfrisby** [comment]
OK. I addressed each comment with a commit, and now I'm going to squash those down and then `bors` it.
